odl/solvers/functional/default_functionals.py

```diff
@@ -570,6 +573,465 @@ def __init__(self, space):
+        """Return the KL-diveregnce in the point ``x``.
+
+        If any components of ``x`` is non-positive, the value is positive
+        infinity.
+        """
+        if self.prior is None:
+            tmp = ((x - 1 - np.log(x)).inner(self.domain.one()))
+        else:
+            tmp = ((x - self.prior + self.prior * np.log(self.prior / x))
+                   .inner(self.domain.one()))
+        if np.isnan(tmp):
+            # In this case, some element was less than or equal to zero
+            return np.inf
+        else:
+            return tmp
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        The gradient is not defined in points where one or more components
+        are non-positive.
+        """
+        functional = self
+
+        class KLGradient(Operator):
+
+            """The gradient operator of this functional."""
+
+            def __init__(self):
+                """Initialize a new instance."""
+                super().__init__(functional.domain, functional.domain,
+                                 linear=False)
+
+            def _call(self, x):
+                """Apply the gradient operator to the given point.
+
+                The gradient is not defined in points where one or more
+                components are non-positive.
+                """
+                if functional.prior is None:
+                    return (-1.0) / x + 1
+                else:
+                    return (-functional.prior) / x + 1
+
+        return KLGradient()
+
+    @property
+    def proximal(self):
+        """Return the proximal factory of the functional.
+
+        See Also
+        --------
+        proximal_cconj_kl : proximal factory for convex conjugate of KL.
+        proximal_cconj : proximal of the convex conjugate of a functional
+        """
+        return proximal_cconj(proximal_cconj_kl(space=self.domain,
+                                                g=self.prior))
+
+    @property
+    def convex_conj(self):
+        """The convex conjugate functional of the KL-functional."""
+        return KullbackLeiblerConvexConj(self.domain, self.prior)
+
+
+class KullbackLeiblerConvexConj(Functional):
+
+    """The convex conjugate of Kullback-Leibler divergence functional.
+
+    Notes
+    -----
+    The functional is given by
+
+    .. math::
+        \\sum_i \left(-g_i \ln(pos({1_X}_i - x_i)) \\right) +
+        I_{1_X - x \geq 0}(x)
+
+    where :math:`g` is the prior, and :math:`I_{1_X - x \geq 0}(x)` is the
+    indicator function on :math:`x \leq 1`.
+
+    This is the convex conjugate functional of `KullbackLeibler`.
+
+    See Also
+    --------
+    KullbackLeibler : convex conjugate functional
+    """
+
+    def __init__(self, space, prior=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp` or `FnBase`
+            Domain of the functional.
+        g : ``space`` element, optional
```

Comment:
Might as well go for 

```
``space`` `element-like`
```

and call `self.__prior = self.domain.element(prior)` further down, instead of the "if prior in space" thing.